x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -672,6 +677,33 @@ public List<Alias> visitFields(EsqlBaseParser.FieldsContext ctx) {
         return ctx != null ? visitList(this, ctx.field(), Alias.class) : new ArrayList<>();
     }
 
+    @Override
+    public NamedExpression visitAggField(EsqlBaseParser.AggFieldContext ctx) {
+        Source source = source(ctx);
+        Alias field = visitField(ctx.field(), source);
+        var filterExpression = ctx.booleanExpression();
+        Expression condition = filterExpression != null ? expression(filterExpression) : null;
+
+        if (condition != null) {
+            Expression child = field.child();
+            // basic check as the filter can be specified only on a function (should be an aggregate but we can't determine that yet)
+            if (field.child().anyMatch(Function.class::isInstance)) {
+                field = field.replaceChild(new FilteredExpression(source, child, condition));
+            }
+            // allow condition only per aggregated function
+            else {
+                source = source(filterExpression);
+                throw new ParsingException(source, "WHERE clause allowed only for aggregate functions [{}]", source.text());
```

Comment:
I've updated this so the aggregate function and the filter is returned with the error pointing to the filter location.
That's because a STATS can be quite large if multiple aggregations are specified.